### PR TITLE
Add shell completion script generation to CLI applications

### LINKED ISSUE #60: Support for terminal/CLI auto-completion (e.g. for fish, zsh, ...)
Given we're already expressing the CLI "schema" (options, arguments, commands etc) in a structured way, we could leverage this information as foundation for CLI auto-completion e.g. for fish/zsh.

**Examples/related links**:
- https://fishshell.com/docs/current/completions.html
- https://github.com/fish-shell/fish-shell/blob/master/share/completions/yarn.fish
- https://docs.npmjs.com/cli/v7/commands/npm-completion
  image

### PR DESCRIPTION
**Pull request by IMax153**
This PR attempts to add support for generation of shell completions to `@effect-ts/cli`. 

The initial goal will be to support `bash` and `zsh` shells, with support for other shells planned in a future feature release. In addition, custom completions are also a goal for a future feature release.

Closes #60.

### CODE REVIEW

**Comment by IMax153** [comment]
@schickling - would love your feedback on the `registerCompletion` combinator I added to allow users to override completions. See the tests here: https://github.com/Effect-TS/cli/blob/d188114c5b0c3f99abaaaaaf6e1d90d7c8977cdd/packages/cli/test/completions.test.ts#L141

File: packages/cli/test/completions.test.ts [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+  })
+
+  describe("Custom", () => {
+    describe("Bash", () => {
+      it("should register custom bash completions for a command", () =>
+        T.gen(function* (_) {
+          const result = yield* _(
+            pipe(
+              gitCommand,
+              Command.registerCompletion(() => T.succeed(Set.singleton("--foo"))),
```
Comment (schickling) [inline packages/cli/test/completions.test.ts:130]: Why does `--foo` have to be a `Set.singleton`?

File: packages/cli/test/completions.test.ts [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+          expect(result).toEqual(Set.fromArray(Equal.string)(["--bash", "--foo"]))
+        }))
+
+      it("should register custom bash completions for an option", () =>
+        T.gen(function* (_) {
+          const fooOption = pipe(
+            Options.text("foo"),
+            Options.withDescription("the foo option"),
+            Options.registerCompletion(() => T.succeed(Set.singleton("--bar")))
```
Comment (schickling) [inline packages/cli/test/completions.test.ts:146]: I don't quite get the relationship of option `foo` with `--bar`?

File: packages/cli/test/completions.test.ts (hunk 1 above)
Comment (IMax153) [reply packages/cli/test/completions.test.ts]: As discussed on our call, a `Set<string>` must be returned to deduplicate completion strings when output to the shell.

File: packages/cli/test/completions.test.ts (hunk 2 above)
Comment (IMax153) [reply packages/cli/test/completions.test.ts]: This test is demonstrating the fact that custom completions override default completions. For example, this option would normally complete as `--foo`, but with a custom completion it now completes as `--bar`. I adjusted the combinator documentation to reflect this.

**Pull request merged**